### ISSUE TITLE
Faster git clones when the cache has a clone of the repo with Gith 2.3

### ISSUE THREAD

**Issue by stof** [composer/composer#3722]
when the composer cache has a clone of the repo already (when using the ``git`` driver), Composer could use the ``--dissociate --reference path/to/git/clone`` option to make the clone faster (this was not safe for older versions of Git because hard references require the referenced repo to keep existing but the new ``--dissociate`` option solves this)

**Comment by AD7six** [comment]
:+1: 

That aught to greatly speed up cloning repeatedly-used dev dependencies.

**Comment by burzum** [comment]
:+1:

**Comment by netstyler** [comment]
:+1:

**Comment by stof** [comment]
@AD7six for anything coming from github, they are not in the composer cache currently, so it would not help much in this case (we would either have to clone in the cache first to create it, or to use another local project as reference but we don't know where they are)

**Comment by AD7six** [comment]
@stof oh? Would it not make sense for dev dependencies to:

 * Use the cache
 * Always update the cache first
 * create dev dependencies using `--dissociate --reference cache` (or even, just clone directly from the cache)

?

I can take a look at implementing that, would also open the door to allow composer to work offline... which may or may not be useful, depending on how often one develops whilst travelling.

**Comment by hkdobrev** [comment]
I think either this or #1323 should be closed in favour of the other.

**Comment by PhilLello** [comment]
I'm working on a patch for this at the moment, should be available at https://github.com/dunlop-lello/composer over the next few days. I haven't used --dissociate at the moment; tbh I'm not convinced there's a need, since people should be pushing their repos off-machine anyway if they care about error recovery, and could simply re-download. That said, I'm open to persuasion, but the horror stories I've read all seem to be from people who didn't have an off-machine backup of their repo.

**Comment by PhilLello** [comment]
WIP Patch:

```
index 94a8d78..df2e200 100644
--- a/src/Composer/Downloader/GitDownloader.php
+++ b/src/Composer/Downloader/GitDownloader.php
@@ -42,14 +42,28 @@ public function doDownload(PackageInterface $package, $path, $url)
     {
         GitUtil::cleanEnv();
         $path = $this->normalizePath($path);
+        $cachePath = $this->config->get('cache-vcs-dir') . '/' . preg_replace('{[^a-z0-9.]}i', '-', $url);
 
         $ref = $package->getSourceReference();
         $flag = Platform::isWindows() ? '/D ' : '';
-        $command = 'git clone --no-checkout %s %s && cd '.$flag.'%2$s && git remote add composer %1$s && git fetch composer';
+
+        if (!file_exists($cachePath))
+        {
+            $command = 'git clone --mirror %s %s';
+            $this->io->writeError("    Caching ".$url);
+
+            $commandCallable = function ($url) use ($cachePath, $command) {
+                return sprintf($command, ProcessExecutor::escape($url), ProcessExecutor::escape($cachePath));
+            };
+
+            $this->gitUtil->runCommand($commandCallable, $url, $path, true);
+        }
+
+        $command = 'git clone --reference %s --no-checkout %s %s && cd '.$flag.'%3$s && git remote add composer %2$s && git fetch composer';
         $this->io->writeError("    Cloning ".$ref);
 
-        $commandCallable = function ($url) use ($ref, $path, $command) {
-            return sprintf($command, ProcessExecutor::escape($url), ProcessExecutor::escape($path), ProcessExecutor::escape($ref));
+        $commandCallable = function ($url) use ($cachePath, $ref, $path, $command) {
+            return sprintf($command, ProcessExecutor::escape($cachePath), ProcessExecutor::escape($url), ProcessExecutor::escape($path), ProcessExecutor::escape($ref));
         };
 
         $this->gitUtil->runCommand($commandCallable, $url, $path, true);

```

**Comment by stof** [comment]
@PhilLello please open a pull request rather than pasting patches in the comment. It is easier to review (we can ask github for more context), it runs the tests on CI and it allows merging easily

**Comment by stof** [comment]
btw, there is already a PR working on it: https://github.com/composer/composer/pull/5384

**Comment by Seldaek** [comment]
Fixed by #5384
